libstuff/libstuff.cpp

```diff
@@ -142,6 +157,58 @@ vector<string> SException::details() const noexcept {
     return stack;
 }
 
+void SSyslogSocketDirect(int priority, const char *format, ...) {
+    int socketError = 0;
+    static const size_t MAX_MESSAGE_SIZE = 8 * 1024;
+
+    // This block shouldn't be required, but it kicks in if this is used from somewhere that fails to call
+    // `SInitialize`. I figured this is better than just losing the log.
+    if (SSyslogSocketFD == 0) {
+        SSyslogSocketAddr.sun_family = AF_UNIX;
+        strcpy(SSyslogSocketAddr.sun_path, "/run/systemd/journal/syslog");
+        SSyslogSocketFD = -1;
+    }
+
+    if (SSyslogSocketFD == -1) {
+        // Create a socket if there isn't one already.
+        SSyslogSocketFD = socket(AF_UNIX, SOCK_DGRAM, 0);
+    }
+    if (SSyslogSocketFD == -1) {
+        // Error opening the socket. We'll fall back to regular syslog.
+        socketError = errno;
+    } else {
+        // At this point, the socket should be good.
+
+        // This is based on the message format here: https://tools.ietf.org/html/rfc5424#section-6 but doesn't actually
+        // match the format there, as it seems syslog ignores almost the entire "header" and "structured data" fields
+        // and treats them as part of the message.
+        string messageHeader = "<" + to_string(8 + priority) + ">" + SProcessName + ": ";
+        thread_local char messageBuffer[MAX_MESSAGE_SIZE];
+        strcpy(messageBuffer, messageHeader.c_str());
+        va_list argptr;
+        va_start(argptr, format);
+        int bytesWritten = vsnprintf(messageBuffer + messageHeader.size(), MAX_MESSAGE_SIZE - messageHeader.size(), format, argptr);
+        va_end(argptr);
+
+
+        // Assume we send the whole message. We don't do anything to handle message truncation.
+        ssize_t bytesSent = sendto(SSyslogSocketFD, messageBuffer, bytesWritten + messageHeader.size(), 0, (struct sockaddr *)&SSyslogSocketAddr, sizeof(struct sockaddr_un));
+        if (bytesSent == -1) {
+            socketError = errno;
+            close(SSyslogSocketFD);
+            SSyslogSocketFD = -1;
+        }
+    }
+
+    if (socketError) {
```

Comment:
NAB In the unlikely scenario something is wrong and every message errors, is there a point at which we'd want to switch `SSyslogFunc` back to its original value and stop trying to use the direct logger?

Reply:
Maybe. I don’t think we need an automated solution for this at this point, though.